neo4j/internal/bolt/connect.go

```
@@ -115,3 +134,181 @@ func Connect(ctx context.Context,
 	}
 	return boltConn, nil
 }
+
+// performManifestNegotiation handles the manifest-style handshake.
+// Returns the negotiated protocol's major and minor version.
+func performManifestNegotiation(
+	ctx context.Context,
+	conn io.ReadWriteCloser,
+	serverName string,
+	errorListener ConnectionErrorListener,
+	boltLogger log.BoltLogger,
+	response []byte,
+) (byte, byte, error) {
+	reader := racing.NewRacingReader(conn)
+
+	// Read the protocol offerings.
+	count, supported, err := readProtocolOfferings(ctx, reader, serverName, errorListener)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	// Read the capability mask.
+	_, capBytes, err := readCapabilityMask(ctx, reader, serverName, errorListener)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	// Log the complete server handshake message.
+	logManifestHandshake(boltLogger, response, count, supported, capBytes)
+
+	// Select an acceptable protocol version.
+	chosen, err := selectProtocol(supported, errorListener, serverName)
+	if err != nil {
+		invalidHandshake := []byte{0x00, 0x00, 0x00, 0x00, 0x00} // 4 bytes for version + 1 byte for capabilities.
+		if _, err := conn.Write(invalidHandshake); err != nil {
+			errorListener.OnDialError(ctx, serverName, err)
+		}
+		return 0, 0, err
```

`selectProtocol` does not call `errorListener.OnDialError` on error. This place only calls it if writing `invalidHandshake` errors out. So if `selectProtocol` failes, but `conn.Write(invalidHandshake)` doesn't for what ever reason, we don't call `OnDialError`. I think `conn.Write(invalidHandshake)` failing can be silently ignored. It's just a best-effort to tell the server "yeah no :point_right::point_left: listen... I don't think this us-thing works, sorry, it's me, not you :broken_heart:". If the server doesn't receive it, just closing the connection and returning the error is fine here. However, `selectProtocol` should be calling `OnDialError` when causing an error.

---

makes sense, refactored 👍